projects/seeker/agents/gpt2_seeker_modules.py

```diff
@@ -0,0 +1,553 @@
+import torch
+import torch.nn
+
+from parlai.agents.fid.fid import FidModel, concat_enc_outs
+from parlai.agents.hugging_face.gpt2 import GPT2Decoder, Gpt2Agent
+from parlai.agents.rag.args import RetrieverType
+from parlai.agents.rag.modules import RagModel
+from parlai.agents.rag.retrievers import (
+    Document,
+    BLANK_DOC,
+    SearchQuerySearchEngineRetriever,
+    retriever_factory,
+)
+from parlai.core.dict import DictionaryAgent
+from parlai.core.opt import Opt
+import parlai.utils.logging as logging
+
+from projects.seeker.agents.seeker_modules import (
+    ComboFidSearchQuerySearchEngineRetriever,
+)
+from projects.seeker.agents.seeker_modules import (
+    combo_fid_retriever_factory,
+    interleave_fid_combo_outputs,
+    ComboFidModel,
+)
+
+########################
+# Retrieval Components #
+########################
+
+
+class FilterDocsForLabelSearchEngineRetrieverMixin:
+    """
+    Mixin for filtering the document containing the label out from retrieved docs.
+
+    As these models are generally used with a CC knowledge dump, and the task on which
+    they are evaluated is from the same CC dump, it is useful to filter out the document
+    containing the label string from the retrieved documents, for a fair comparison.
+    """
+
+    def set_labels(self, labels: torch.Tensor):
+        """
+        Cache the label vec.
+        """
+        self.label_vec = labels
+
+    def retrieve_and_score(
+        self, query: torch.LongTensor
+    ) -> Tuple[List[List[Document]], torch.Tensor]:
+        """
+        Override retrieve and score to filter out docs that contain the label string.
+
+        Copy over the whole thing because we need to check before chunking.
+        """
+        # step 1
+        search_queries = self.generate_search_query(query)  # type: ignore
+
+        # step 2
+        search_results_batch = self.search_client.retrieve(
+            search_queries, self.n_docs
+        )  # type: ignore
+
+        # step 3
+        top_docs = []
+        top_doc_scores = []
+        max_n_docs: int = self.n_docs  # type: ignore
+        for batch_id, (sq, search_results) in enumerate(
+            zip(search_queries, search_results_batch)
+        ):
+            if not search_results:
+                search_results = self._empty_docs(self.n_docs)  # type: ignore
+            elif len(search_results) < self.n_docs:  # type: ignore
+                remain_docs = self.n_docs - len(search_results)  # type: ignore
+                search_results.extend(self._empty_docs(remain_docs))  # type: ignore
+            docs_i = []
+            scors_i = []
+            # Change this debug later
+            logging.debug(
+                f'URLS:\n{self._display_urls(search_results)}'
+            )  # type: ignore
+            label_text = self.dict.vec2txt(
+                self.label_vec[batch_id, :-1]
+            )  # type: ignore
+            for i, doc in enumerate(search_results):
+                url = doc['url']
```

Comment:
nit: thinking about the variation of the search module used, often they may not return or title. And the KeyError generated here may not be very informative for someone who uses it but doesn't know it's internal. How about checking if all these exist and having an assert that tells user what was expected, or a load warning and then replacing the missing entries with a deafult?

Reply:
that's not a bad idea; I took this function literally verbatim from the one in `parlai/agents/rag/retrievers.py`; the only change is line 104 to 106. if we want to change that we should go directly to the function there